### ISSUE TITLE
Make P2 repo URL configurable

### ISSUE THREAD

**Issue by jhonnen** [diffplug/spotless#1629]
`spotless-plugin-gradle:6.17.0` broke our build because it now tries to download artifacts directly from `https://download.eclipse.org/eclipse/updates` whereas before they were resolved via the configured repositories in Gradle (i.e. an internal maven-central mirror).

Please make the P2 repo URL which is currently hardcoded in EquoBasedStepBuilder#addPlatformRepo configurable so we can swap it out with an internal mirror.

Stacktrace:

	Caused by: java.lang.RuntimeException: java.net.SocketTimeoutException: connect timed out
		at dev.equo.solstice.p2.Unchecked.wrap(Unchecked.java:25)
		at dev.equo.solstice.p2.P2Model.query(P2Model.java:133)
		at com.diffplug.spotless.extra.EquoBasedStepBuilder.get(EquoBasedStepBuilder.java:88)
		at com.diffplug.spotless.FormatterStepImpl.calculateState(FormatterStepImpl.java:58)
		at com.diffplug.spotless.LazyForwardingEquality.state(LazyForwardingEquality.java:56)
		at com.diffplug.spotless.LazyForwardingEquality.unlazy(LazyForwardingEquality.java:118)
		at com.diffplug.spotless.LazyForwardingEquality.unlazy(LazyForwardingEquality.java:124)
		at com.diffplug.gradle.spotless.SpotlessExtensionPredeclare.lambda$new$0(SpotlessExtensionPredeclare.java:39)
		at com.diffplug.gradle.spotless.SpotlessExtensionPredeclare.lambda$new$1(SpotlessExtensionPredeclare.java:33)
	[...]
	Caused by: java.net.SocketTimeoutException: connect timed out
		at okhttp3.internal.platform.Platform.connectSocket(Platform.kt:128)
		at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:295)
		at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:207)
		at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:226)
		at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106)
		at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74)
		at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255)
		at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
		at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
		at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
		at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
		at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
		at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
		at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
		at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
		at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
		at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
		at dev.equo.solstice.p2.P2Client.getBytes(P2Client.java:145)
		at dev.equo.solstice.p2.P2Client.getString(P2Client.java:116)
		at dev.equo.solstice.p2.P2Client$Folder.<init>(P2Client.java:209)
		at dev.equo.solstice.p2.P2Client.addUnits(P2Client.java:94)
		at dev.equo.solstice.p2.P2Session.populateFrom(P2Session.java:34)
		at dev.equo.solstice.p2.P2Model.queryRaw(P2Model.java:96)
		at dev.equo.solstice.p2.P2Model.query(P2Model.java:125)
		... 191 more

**Comment by nedtwigg** [comment]
Happy to merge a PR for this! I'm open to other ideas, but my instinct is to do it like this:

Add a method `replaceAllP2ReposWith(String...)` to `EquoBasedStepBuilder`

Then right here

https://github.com/diffplug/spotless/blob/a92c7b04be4e2d1a896555c302505b903deae2fe/lib-extra/src/main/java/com/diffplug/spotless/extra/EquoBasedStepBuilder.java#L87-L88

you can do 

```java
var model = model(formatterVersion)
if (replaceAllP2Repos != null) {
  model.getP2repo().removeAll()
  model.getP2repo().addAll(replaceAllP2Repos)
}
```

It'll also have to get wired into the Gradle and Maven DSL. Happy to merge a PR which only addresses one step and one DSL, you don't have to solve the entire problem if you don't want to, just your own.

**Comment by jhonnen** [comment]
> Add a method `replaceAllP2ReposWith(String...)` to `EquoBasedStepBuilder`

But this would quietly override the configured version, right? Just overriding the `"https://download.eclipse.org/eclipse/updates/"` part would cause less surprises IMHO.
But that only takes care of JDT, the other two are adding additional repositories...

How about something like this: 

	Map<String, String> p2Mirrors;

	/** Creates the state of the configuration. */
	EquoBasedStepBuilder.State get() throws Exception {
		P2Model model = model(formatterVersion);
		ArrayList<String> p2Repos = new ArrayList<>(model.getP2repo());
		p2Repos.replaceAll(url -> {
			for (Map.Entry<String, String> mirror : p2Mirrors.entrySet()) {
				if (url.startsWith(mirror.getKey())) {
					return mirror.getValue() + url.substring(mirror.getKey().length());
				}
			}
			return url;
		});
		model.getP2repo().addAll(p2Repos);

		var query = model.query(P2ClientCache.PREFER_OFFLINE, P2QueryCache.ALLOW);
		// ...


And then you would configure mirrors somewhere (not sure where exactly...):

    spotless {
        p2Mirrors = [
            'https://download.eclipse.org/eclipse/updates': 'https://some-internal-mirror/eclipse-updates',
            'https://groovy.jfrog.io/artifactory/plugins-release': '...',
        ]
    }

**Comment by nedtwigg** [comment]
That's a great idea! I have two notes:

1. if any repos are being replaced, all of them should be. User should get an error message "you need to specify a replacement repo for blah"
2. I would do the replacement like this `jdt().replaceAllP2ReposByPrefix(['https://download.eclipse.org/eclipse/updates' : 'https://internal'])`

**Comment by nedtwigg** [comment]
Published in `plugin-gradle 6.18.0`. Thanks to @jhonnen!
